{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Using SlimSuite to find SLiMs in different peptide groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading data...\n",
      "\n",
      "Filtering Data \n",
      "Number of peptides in full library (VT/VP): 28112\n",
      "Number of peptides used in RITA experiment (VT/VP): 25735\n",
      "\n",
      "--- Starting SLiMFinder Analysis for Peptide Groups ---\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "74664e9c40f9469f81e21987c6b4762e",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Processing peptide groups for SLiMFinder:   0%|          | 0/1 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generating FASTA for 10 peptides...\n",
      "Generated FASTA file: /scratch/382110.1.ood/tmpii9my36e_Experiment_Upregulated_VT_VP.fasta with 10 sequences.\n",
      "\n",
      "--- Running SLiMFinder for group: Experiment_Upregulated_VT_VP ---\n",
      "  Using minocc_perc=0.1 for 10 peptides, resulting in absolute minocc=2\n",
      "SLiMFinder command: python /projectnb/cancergrp/Philipp/SLiMSuite/tools/slimfinder.py seqin=/scratch/382110.1.ood/tmpii9my36e_Experiment_Upregulated_VT_VP.fasta resdir=/projectnb/cancergrp/Philipp/results/RITA_peptides/slimfinder_analyses/slimfinder_output_Experiment_Upregulated_VT_VP/ resfile=/projectnb/cancergrp/Philipp/results/RITA_peptides/slimfinder_analyses/slimfinder_output_Experiment_Upregulated_VT_VP/Experiment_Upregulated_VT_VP_slimfinder_results.csv efilter=F masking=F slimlen=10 maxwild=2 minocc=2 probcut=0.05 maxseq=0 v=0 runid=Experiment_Upregulated_VT_VP forks=1\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[11], line 219\u001b[0m\n\u001b[1;32m    216\u001b[0m create_fasta_from_df(group_df, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124midentifier\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mAminoacids\u001b[39m\u001b[38;5;124m'\u001b[39m, fasta_input_path)\n\u001b[1;32m    218\u001b[0m \u001b[38;5;66;03m# Run SLiMFinder with chosen parameters\u001b[39;00m\n\u001b[0;32m--> 219\u001b[0m group_slim_results \u001b[38;5;241m=\u001b[39m \u001b[43mrun_slimfinder\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    220\u001b[0m \u001b[43m    \u001b[49m\u001b[43mgroup_name\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mgroup_name\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    221\u001b[0m \u001b[43m    \u001b[49m\u001b[43minput_fasta_path\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mfasta_input_path\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    222\u001b[0m \u001b[43m    \u001b[49m\u001b[43moutput_base_dir\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mslimfinder_output_base_dir\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    223\u001b[0m \u001b[43m    \u001b[49m\u001b[43mslimfinder_script\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mSLIMFINDER_SCRIPT\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    224\u001b[0m \u001b[43m    \u001b[49m\u001b[43mpython_executable\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mPYTHON_EXECUTABLE\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    225\u001b[0m \u001b[43m    \u001b[49m\u001b[43mslimlen\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m10\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m     \u001b[49m\u001b[38;5;66;43;03m# Max SLiM length\u001b[39;49;00m\n\u001b[1;32m    226\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmaxwild\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m2\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m      \u001b[49m\u001b[38;5;66;43;03m# Max wildcards (e.g., A.B.C)\u001b[39;49;00m\n\u001b[1;32m    227\u001b[0m \u001b[43m    \u001b[49m\u001b[43mminocc_perc\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m0.1\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;66;43;03m# Motif must occur in at least 3 unique peptides\u001b[39;49;00m\n\u001b[1;32m    228\u001b[0m \u001b[43m    \u001b[49m\u001b[43mprobcut\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m0.05\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m   \u001b[49m\u001b[38;5;66;43;03m# p-value cutoff for significance (higher allows more motifs)\u001b[39;49;00m\n\u001b[1;32m    229\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmaxseq\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m       \u001b[49m\u001b[38;5;66;43;03m# 0 means no limit on input sequences (peptides)\u001b[39;49;00m\n\u001b[1;32m    230\u001b[0m \u001b[43m    \u001b[49m\u001b[43mnr_peptides\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;28;43mlen\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mgroup_df\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;66;43;03m# Number of peptides in group\u001b[39;49;00m\n\u001b[1;32m    231\u001b[0m \u001b[43m    \u001b[49m\u001b[43mnum_forks\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mNUM_FORKS_SLIMFINDER\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;66;43;03m# Number of forks to create (number of CPUs available)\u001b[39;49;00m\n\u001b[1;32m    232\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    233\u001b[0m slimfinder_results_summary[group_name] \u001b[38;5;241m=\u001b[39m group_slim_results\n\u001b[1;32m    235\u001b[0m \u001b[38;5;66;03m# Clean up the temporary FASTA file\u001b[39;00m\n",
      "Cell \u001b[0;32mIn[11], line 134\u001b[0m, in \u001b[0;36mrun_slimfinder\u001b[0;34m(group_name, input_fasta_path, output_base_dir, slimfinder_script, python_executable, slimlen, maxwild, minocc_perc, probcut, maxseq, nr_peptides, num_forks)\u001b[0m\n\u001b[1;32m    129\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mSLiMFinder command: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;241m.\u001b[39mjoin(command)\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m    131\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    132\u001b[0m     \u001b[38;5;66;03m# Execute the command, capturing stdout/stderr. check=True raises error for non-zero exit codes.\u001b[39;00m\n\u001b[1;32m    133\u001b[0m     \u001b[38;5;66;03m# Assuming the environment (conda and modules) is activated *before* running Python.\u001b[39;00m\n\u001b[0;32m--> 134\u001b[0m     process \u001b[38;5;241m=\u001b[39m \u001b[43msubprocess\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcommand\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcheck\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcapture_output\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtext\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m)\u001b[49m\n\u001b[1;32m    135\u001b[0m     \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mSLiMFinder for \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mgroup_name\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m completed successfully.\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m    136\u001b[0m     \u001b[38;5;66;03m# Uncomment these lines for detailed debug output from SLiMFinder\u001b[39;00m\n\u001b[1;32m    137\u001b[0m     \u001b[38;5;66;03m# print(\"SLiMFinder Stdout:\\n\", process.stdout) \u001b[39;00m\n\u001b[1;32m    138\u001b[0m     \u001b[38;5;66;03m# print(\"SLiMFinder Stderr:\\n\", process.stderr) \u001b[39;00m\n\u001b[1;32m    139\u001b[0m \n\u001b[1;32m    140\u001b[0m     \u001b[38;5;66;03m# Read the main results CSV\u001b[39;00m\n",
      "File \u001b[0;32m/projectnb/cancergrp/Philipp/.conda/envs/slimsuite_env/lib/python3.9/subprocess.py:507\u001b[0m, in \u001b[0;36mrun\u001b[0;34m(input, capture_output, timeout, check, *popenargs, **kwargs)\u001b[0m\n\u001b[1;32m    505\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m Popen(\u001b[38;5;241m*\u001b[39mpopenargs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs) \u001b[38;5;28;01mas\u001b[39;00m process:\n\u001b[1;32m    506\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 507\u001b[0m         stdout, stderr \u001b[38;5;241m=\u001b[39m \u001b[43mprocess\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcommunicate\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43minput\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeout\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    508\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m TimeoutExpired \u001b[38;5;28;01mas\u001b[39;00m exc:\n\u001b[1;32m    509\u001b[0m         process\u001b[38;5;241m.\u001b[39mkill()\n",
      "File \u001b[0;32m/projectnb/cancergrp/Philipp/.conda/envs/slimsuite_env/lib/python3.9/subprocess.py:1134\u001b[0m, in \u001b[0;36mPopen.communicate\u001b[0;34m(self, input, timeout)\u001b[0m\n\u001b[1;32m   1131\u001b[0m     endtime \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m   1133\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 1134\u001b[0m     stdout, stderr \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_communicate\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43minput\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mendtime\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1135\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyboardInterrupt\u001b[39;00m:\n\u001b[1;32m   1136\u001b[0m     \u001b[38;5;66;03m# https://bugs.python.org/issue25942\u001b[39;00m\n\u001b[1;32m   1137\u001b[0m     \u001b[38;5;66;03m# See the detailed comment in .wait().\u001b[39;00m\n\u001b[1;32m   1138\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m timeout \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n",
      "File \u001b[0;32m/projectnb/cancergrp/Philipp/.conda/envs/slimsuite_env/lib/python3.9/subprocess.py:1995\u001b[0m, in \u001b[0;36mPopen._communicate\u001b[0;34m(self, input, endtime, orig_timeout)\u001b[0m\n\u001b[1;32m   1988\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_check_timeout(endtime, orig_timeout,\n\u001b[1;32m   1989\u001b[0m                         stdout, stderr,\n\u001b[1;32m   1990\u001b[0m                         skip_check_and_raise\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[1;32m   1991\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mRuntimeError\u001b[39;00m(  \u001b[38;5;66;03m# Impossible :)\u001b[39;00m\n\u001b[1;32m   1992\u001b[0m         \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m_check_timeout(..., skip_check_and_raise=True) \u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[1;32m   1993\u001b[0m         \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mfailed to raise TimeoutExpired.\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[0;32m-> 1995\u001b[0m ready \u001b[38;5;241m=\u001b[39m \u001b[43mselector\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mselect\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtimeout\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1996\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_check_timeout(endtime, orig_timeout, stdout, stderr)\n\u001b[1;32m   1998\u001b[0m \u001b[38;5;66;03m# XXX Rewrite these to use non-blocking I/O on the file\u001b[39;00m\n\u001b[1;32m   1999\u001b[0m \u001b[38;5;66;03m# objects; they are no longer using C stdio!\u001b[39;00m\n",
      "File \u001b[0;32m/projectnb/cancergrp/Philipp/.conda/envs/slimsuite_env/lib/python3.9/selectors.py:416\u001b[0m, in \u001b[0;36m_PollLikeSelector.select\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    414\u001b[0m ready \u001b[38;5;241m=\u001b[39m []\n\u001b[1;32m    415\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 416\u001b[0m     fd_event_list \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_selector\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpoll\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtimeout\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    417\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mInterruptedError\u001b[39;00m:\n\u001b[1;32m    418\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m ready\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "import sys\n",
    "import os\n",
    "import subprocess\n",
    "import tempfile\n",
    "import csv\n",
    "import pandas as pd\n",
    "from collections import Counter\n",
    "from tqdm.auto import tqdm\n",
    "import argparse\n",
    "import openpyxl\n",
    "import math\n",
    "\n",
    "#parser = argparse.ArgumentParser(description=\"Find SLiMs for different peptide groups.\")\n",
    "#parser.add_argument(\"--num_forks\", type=int, default=1,\n",
    "#                    help=\"Number of CPU forks for SLiMFinder parallel execution.\")\n",
    "#args = parser.parse_args()\n",
    "\n",
    "#NUM_FORKS_SLIMFINDER = args.num_forks\n",
    "NUM_FORKS_SLIMFINDER = 1\n",
    "\n",
    "# --- SLiMSuite Configuration ---\n",
    "# IMPORTANT: Adjust this path to where you cloned SLiMSuite\n",
    "SLIMSUITE_PATH = \"/projectnb/cancergrp/Philipp/SLiMSuite\" \n",
    "SLIMFINDER_SCRIPT = os.path.join(SLIMSUITE_PATH, \"tools\", \"slimfinder.py\")\n",
    "\n",
    "# Ensure the SLiMFinder script exists\n",
    "if not os.path.exists(SLIMFINDER_SCRIPT):\n",
    "    print(f\"ERROR: SLiMFinder script not found at {SLIMFINDER_SCRIPT}\")\n",
    "    print(\"Please ensure SLiMSuite is cloned and SLIMSUITE_PATH is set correctly.\")\n",
    "    sys.exit(1)\n",
    "\n",
    "# Ensure this points to the Python executable within your 'slimsuite_env'\n",
    "# When you activate your conda environment and run the script, 'python' usually points to it.\n",
    "PYTHON_EXECUTABLE = \"python\" \n",
    "\n",
    "# --- The existing data and results directories ---\n",
    "DATA_DIR = \"/projectnb/cancergrp/Philipp/data/\"\n",
    "RESULTS_DIR = \"/projectnb/cancergrp/Philipp/results/RITA_peptides\"\n",
    "os.makedirs(RESULTS_DIR, exist_ok=True)\n",
    "\n",
    "# --- Load your DataFrames ---\n",
    "print(\"Loading data...\")\n",
    "full_library_df = pd.read_csv(f\"{DATA_DIR}VP_library_all_sequences.csv\")\n",
    "RITA_exp_df = pd.read_excel(f\"{DATA_DIR}RITA_and_ABT_pos_selection_screens.xlsx\", sheet_name='RITA')\n",
    "\n",
    "\n",
    "\n",
    "# --- SLiMSuite Input/Output Helper Functions ---\n",
    "\n",
    "def create_fasta_from_df(df, id_col, seq_col, output_fasta_path):\n",
    "    \"\"\"\n",
    "    Creates a FASTA file from a DataFrame, ensuring valid headers and sequences.\n",
    "    Args:\n",
    "        df (pd.DataFrame): Input DataFrame.\n",
    "        id_col (str): Name of the column containing identifiers.\n",
    "        seq_col (str): Name of the column containing amino acid sequences.\n",
    "        output_fasta_path (str): Path for the output FASTA file.\n",
    "    \"\"\"\n",
    "    print(f\"Generating FASTA for {len(df)} peptides...\")\n",
    "    with open(output_fasta_path, 'w') as f:\n",
    "        # Filter out rows with missing or non-string identifiers/sequences\n",
    "        valid_rows = df[df[id_col].notna() & df[seq_col].notna()].copy()\n",
    "        \n",
    "        for index, row in valid_rows.iterrows():\n",
    "            identifier = str(row[id_col]).strip()\n",
    "            # Replace spaces in identifier with underscores for FASTA header compatibility\n",
    "            identifier = identifier.replace(' ', '_').replace('|', '_').replace('>', '') \n",
    "            sequence = str(row[seq_col]).strip()\n",
    "            \n",
    "            if identifier and sequence: # Only write non-empty identifiers and sequences\n",
    "                f.write(f\">{identifier}\\n{sequence}\\n\")\n",
    "            else:\n",
    "                # print(f\"Warning: Skipping row {index} due to invalid identifier or sequence: ID='{identifier}', Seq='{sequence[:20]}...'\")\n",
    "                pass # Suppress warning for every skipped row for cleaner output\n",
    "    print(f\"Generated FASTA file: {output_fasta_path} with {len(valid_rows)} sequences.\")\n",
    "\n",
    "def run_slimfinder(group_name, input_fasta_path, output_base_dir, slimfinder_script, python_executable,\n",
    "                   slimlen=10, maxwild=2, minocc_perc=0.1, probcut=0.05, maxseq=0, nr_peptides = 0.1, num_forks=1):\n",
    "    \"\"\"\n",
    "    Runs SLiMFinder for a given group of peptides.\n",
    "    Args:\n",
    "        group_name (str): Name of the peptide group (for output organization).\n",
    "        input_fasta_path (str): Path to the input FASTA file for SLiMFinder.\n",
    "        output_base_dir (str): Base directory for SLiMFinder results.\n",
    "        slimfinder_script (str): Path to the slimfinder.py script.\n",
    "        python_executable (str): Python executable to run slimfinder.py with (e.g., 'python').\n",
    "        slimlen (int): Maximum SLiM length.\n",
    "        maxwild (int): Maximum wildcards between defined positions.\n",
    "        minocc_perc (int): Minimum percentage of occurrences for a motif.\n",
    "        probcut (float): Probability cut-off for returned motifs.\n",
    "        maxseq (int): Maximum number of sequences to process (0 for no limit).\n",
    "        nr_peptides (int): Total number of peptides in this group.\n",
    "    Returns:\n",
    "        pd.DataFrame: A DataFrame of significant motifs found, or empty if none.\n",
    "    \"\"\"\n",
    "    print(f\"\\n--- Running SLiMFinder for group: {group_name} ---\")\n",
    "\n",
    "    # Create a dedicated output directory for this SLiMFinder run\n",
    "    slimfinder_output_dir = os.path.join(output_base_dir, f\"slimfinder_output_{group_name.replace(' ', '_')}\")\n",
    "    os.makedirs(slimfinder_output_dir, exist_ok=True)\n",
    "    \n",
    "    # Calculate absolute minocc based on percentage and total peptides\n",
    "    # Ensure minocc is an integer and at least 2 (a motif must occur at least twice to be a motif) and max the total number of peptides\n",
    "    minocc_abs = max(2, min(int(math.ceil(minocc_perc * nr_peptides)), nr_peptides))\n",
    "\n",
    "    print(f\"  Using minocc_perc={minocc_perc} for {nr_peptides} peptides, resulting in absolute minocc={minocc_abs}\")\n",
    "\n",
    "    # Define the main results file path\n",
    "    results_csv_path = os.path.join(slimfinder_output_dir, f\"{group_name.replace(' ', '_')}_slimfinder_results.csv\")\n",
    "    \n",
    "    # SLiMFinder command arguments\n",
    "    command = [\n",
    "        python_executable, slimfinder_script,\n",
    "        f\"seqin={input_fasta_path}\",\n",
    "        f\"resdir={slimfinder_output_dir}/\", # SLiMFinder expects trailing slash for resdir\n",
    "        f\"resfile={results_csv_path}\",\n",
    "        f\"efilter=F\",      # Turn off evolutionary filtering for short peptides\n",
    "        f\"masking=F\",      # Turn off masking for short peptides\n",
    "        f\"slimlen={slimlen}\",\n",
    "        f\"maxwild={maxwild}\",\n",
    "        f\"minocc={minocc_abs}\",\n",
    "        f\"probcut={probcut}\",\n",
    "        f\"maxseq={maxseq}\", # 0 means no limit on sequences; useful if one has many peptides.\n",
    "        f\"v=0\",             # Verbosity: 0 for default, -1 for silent, 1 for more debug info\n",
    "        f\"runid={group_name.replace(' ', '_')}\", # Add a run ID for easier tracking\n",
    "        f\"forks={num_forks}\" # Forks the task to use several CPUs if provided\n",
    "    ]\n",
    "\n",
    "    print(f\"SLiMFinder command: {' '.join(command)}\")\n",
    "\n",
    "    try:\n",
    "        # Execute the command, capturing stdout/stderr. check=True raises error for non-zero exit codes.\n",
    "        # Assuming the environment (conda and modules) is activated *before* running Python.\n",
    "        process = subprocess.run(command, check=True, capture_output=True, text=True)\n",
    "        print(f\"SLiMFinder for {group_name} completed successfully.\")\n",
    "        # Uncomment these lines for detailed debug output from SLiMFinder\n",
    "        # print(\"SLiMFinder Stdout:\\n\", process.stdout) \n",
    "        # print(\"SLiMFinder Stderr:\\n\", process.stderr) \n",
    "\n",
    "        # Read the main results CSV\n",
    "        if os.path.exists(results_csv_path):\n",
    "            slim_results_df = pd.read_csv(results_csv_path)\n",
    "            print(f\"Found {len(slim_results_df)} motifs for {group_name}.\")\n",
    "            return slim_results_df\n",
    "        else:\n",
    "            print(f\"Warning: SLiMFinder results file not found at {results_csv_path}\")\n",
    "            return pd.DataFrame() # Return empty DataFrame if file not found\n",
    "\n",
    "    except subprocess.CalledProcessError as e:\n",
    "        print(f\"Error running SLiMFinder for {group_name}: {e}\")\n",
    "        print(\"SLiMFinder Stdout:\\n\", e.stdout)\n",
    "        print(\"SLiMFinder Stderr:\\n\", e.stderr)\n",
    "        return pd.DataFrame() # Return empty DataFrame on error\n",
    "    except Exception as e:\n",
    "        print(f\"An unexpected error occurred while running SLiMFinder for {group_name}: {e}\")\n",
    "        return pd.DataFrame() # Return empty DataFrame on error\n",
    "\n",
    "\n",
    "# --- Main Script Logic to filter data and run SLiMFinder ---\n",
    "\n",
    "print(\"\\nFiltering Data \")\n",
    "full_library_filtered = full_library_df[full_library_df['code'].isin(['VT', 'VP'])].copy()\n",
    "RITA_exp_filtered = RITA_exp_df[RITA_exp_df['type'].isin(['VT', 'VP'])].copy()\n",
    "\n",
    "print(f\"Number of peptides in full library (VT/VP): {len(full_library_filtered)}\")\n",
    "print(f\"Number of peptides used in RITA experiment (VT/VP): {len(RITA_exp_filtered)}\")\n",
    "\n",
    "# Ensure 'identifier' is consistent for RITA_exp_filtered\n",
    "RITA_exp_filtered['identifier'] = RITA_exp_filtered['tileID']\n",
    "\n",
    "# Peptides Used vs. Not Used in Experiment (from the VT/VP filtered library)\n",
    "used_sequences_set = set(RITA_exp_filtered['Aminoacids'].unique())\n",
    "not_used_peptides_df = full_library_filtered[~full_library_filtered['Aminoacids'].isin(used_sequences_set)].copy()\n",
    "\n",
    "# Experiment Significant vs. Non-Significant (from VT/VP used in experiment)\n",
    "RITA_sig = RITA_exp_filtered[RITA_exp_filtered['sig'] == 'Yes'].copy()\n",
    "RITA_non_sig = RITA_exp_filtered[RITA_exp_filtered['sig'] == 'No'].copy()\n",
    "\n",
    "# Experiment Upregulated vs. Downregulated Significant (from VT/VP used and significant)\n",
    "RITA_sig['log2FoldChange'] = pd.to_numeric(RITA_sig['log2FoldChange'], errors='coerce')\n",
    "RITA_up = RITA_sig[RITA_sig['log2FoldChange'] > 0].copy()\n",
    "RITA_down = RITA_sig[RITA_sig['log2FoldChange'] < 0].copy()\n",
    "\n",
    "# Store DataFrames in a dictionary for easy iteration and FASTA creation\n",
    "peptide_groups_dfs = {\n",
    "    #'Full_Library_VT_VP': full_library_filtered,\n",
    "    #'Experiment_Used_VT_VP': RITA_exp_filtered,\n",
    "    #'Experiment_Not_Used_VT_VP': not_used_peptides_df,\n",
    "    'Experiment_Significant_VT_VP': RITA_sig,\n",
    "    'Experiment_NonSignificant_VT_VP': RITA_non_sig,\n",
    "    'Experiment_Upregulated_VT_VP': RITA_up,\n",
    "    'Experiment_Downregulated_VT_VP': RITA_down\n",
    "}\n",
    "\n",
    "\n",
    "# --- Run SLiMFinder for each peptide group ---\n",
    "print(\"\\n--- Starting SLiMFinder Analysis for Peptide Groups ---\")\n",
    "slimfinder_results_summary = {}\n",
    "slimfinder_output_base_dir = os.path.join(RESULTS_DIR, \"slimfinder_analyses\")\n",
    "os.makedirs(slimfinder_output_base_dir, exist_ok=True)\n",
    "\n",
    "for group_name, group_df in tqdm(peptide_groups_dfs.items(), desc=\"Processing peptide groups for SLiMFinder\"):\n",
    "    if group_df.empty:\n",
    "        print(f\"Skipping SLiMFinder for empty group: {group_name}\")\n",
    "        slimfinder_results_summary[group_name] = pd.DataFrame()\n",
    "        continue\n",
    "\n",
    "    # Create a temporary FASTA file for SLiMFinder input\n",
    "    # Using tempfile.NamedTemporaryFile to manage temporary files better\n",
    "    with tempfile.NamedTemporaryFile(mode='w', delete=False, suffix=f\"_{group_name.replace(' ', '_')}.fasta\") as tmp_fasta_file:\n",
    "        fasta_input_path = tmp_fasta_file.name\n",
    "    \n",
    "    create_fasta_from_df(group_df, 'identifier', 'Aminoacids', fasta_input_path)\n",
    "\n",
    "    # Run SLiMFinder with chosen parameters\n",
    "    group_slim_results = run_slimfinder(\n",
    "        group_name=group_name,\n",
    "        input_fasta_path=fasta_input_path,\n",
    "        output_base_dir=slimfinder_output_base_dir,\n",
    "        slimfinder_script=SLIMFINDER_SCRIPT,\n",
    "        python_executable=PYTHON_EXECUTABLE,\n",
    "        slimlen=10,     # Max SLiM length\n",
    "        maxwild=2,      # Max wildcards (e.g., A.B.C)\n",
    "        minocc_perc=0.1, # Motif must occur in at least 3 unique peptides\n",
    "        probcut=0.05,   # p-value cutoff for significance (higher allows more motifs)\n",
    "        maxseq=0,       # 0 means no limit on input sequences (peptides)\n",
    "        nr_peptides = len(group_df), # Number of peptides in group\n",
    "        num_forks=NUM_FORKS_SLIMFINDER # Number of forks to create (number of CPUs available)\n",
    "    )\n",
    "    slimfinder_results_summary[group_name] = group_slim_results\n",
    "\n",
    "    # Clean up the temporary FASTA file\n",
    "    if os.path.exists(fasta_input_path):\n",
    "        os.remove(fasta_input_path)\n",
    "\n",
    "print(\"\\n--- SLiMFinder Analysis Complete ---\")\n",
    "print(\"Summary of significant motifs found per group:\")\n",
    "for group, results_df in slimfinder_results_summary.items():\n",
    "    num_motifs = len(results_df)\n",
    "    print(f\"- {group}: {num_motifs} significant motifs\")\n",
    "    if num_motifs > 0:\n",
    "        # Display top 5 motifs (by significance if available, otherwise just first 5)\n",
    "        if 'Sig' in results_df.columns:\n",
    "            print(\"  Top 5 Motifs (by Significance):\")\n",
    "            # Ensure 'Sig' column is numeric, coercing errors to NaN and dropping NaNs before sorting\n",
    "            results_df['Sig'] = pd.to_numeric(results_df['Sig'], errors='coerce')\n",
    "            print(results_df[['Pattern', 'Sig', 'Occ', 'Support']].dropna(subset=['Sig']).sort_values(by='Sig').head())\n",
    "        else:\n",
    "            print(\"  Top 5 Motifs:\")\n",
    "            print(results_df[['Pattern', 'Occ', 'Support']].head())\n",
    "    print(\"-\" * 30)\n",
    "\n",
    "# Optionally, combine all SLiMFinder results into a single DataFrame\n",
    "all_slimfinder_results_combined = pd.concat(\n",
    "    [df.assign(Group=name) for name, df in slimfinder_results_summary.items() if not df.empty]\n",
    ")\n",
    "if not all_slimfinder_results_combined.empty:\n",
    "    combined_slim_results_path = os.path.join(slimfinder_output_base_dir, \"all_groups_slimfinder_combined_results.csv\")\n",
    "    all_slimfinder_results_combined.to_csv(combined_slim_results_path, index=False)\n",
    "    print(f\"\\nAll SLiMFinder results combined and saved to: {combined_slim_results_path}\")\n",
    "\n",
    "print(\"\\nScript execution finished.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6                                            NaN\n",
       "105      RYVEYRDKLVGKTVKVKVIRVDYTKGYIDVNYKRMCRHQ\n",
       "147      QQPSIKDRTKAWEDFLQCTSGESEQVEGGMSKDDGGVER\n",
       "181      QVEGGMSKDDGGVERRSLEDLSSTSPTDGTIGKRVSNTR\n",
       "200      SPTDGTIGKRVSNTRDWAEGSDDIQLDPVVTDVVYHDHG\n",
       "                          ...                   \n",
       "32716    PYSGKREDIWCGSLIGTRARATWAENIQVAINQVRAIIG\n",
       "32717    ENIQVAINQVRAIIGDEKYVDYMSSLKRYEDTTLVEDTV\n",
       "32718    NIQVAINQVRAIIGDEKYVDYMSSLKRYEDTTLVEDTVL\n",
       "32719    VRSYCYLATVSDLSTKAACPTMGDAHNDKRADPAFVCRQ\n",
       "32720    AHNDKRADPAFVCRQGVVDRGWGNGCGLFGKGSIDTCAK\n",
       "Name: Aminoacids, Length: 25735, dtype: object"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "RITA_exp_filtered['Aminoacids'].str[10:]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "jupyter_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
